### PR TITLE
introducing author and date of pause in paused state

### PR DESCRIPTION
**Pull request by eryshev**
BTW, I introduced integration tests of database migrations with doobie scalatest.

### CODE REVIEW

File: core/src/main/scala/com/criteo/cuttle/Executor.scala [hunk 1]
```diff
@@ -128,6 +128,17 @@ class CancellationListener private[cuttle] (execution: Execution[_], private[cut
   def unsubscribeOn[A](f: Future[A]) = f.andThen { case _ => unsubscribe() }(execution.executionContext)
 }
 
+case class PausedJobWithExecutions[S <: Scheduling](id: String,
```
[the remark this answers is not in the document]
Comment (dufrannea) [reply core/src/main/scala/com/criteo/cuttle/Executor.scala]: Sorry but I cannot appear to find a usage of PausedJobWithExecutions where executions is not an empty Map. Could you explain please ?

File: core/src/main/scala/com/criteo/cuttle/Database.scala [hunk 2]
```diff
@@ -85,7 +77,7 @@ private[cuttle] object Database {
     x => x.noSpaces
   )
 
-  val schemaEvolutions = List(
+  private[cuttle] val schemaEvolutions = List(
```
[the remark this answers is not in the document]
Comment (eryshev) [reply core/src/main/scala/com/criteo/cuttle/Database.scala]: ok

File: build.sbt [hunk 3]
```diff
@@ -193,15 +192,16 @@ lazy val cuttle =
       ),
       libraryDependencies ++= Seq(
         "org.tpolecat" %% "doobie-core",
-        "org.tpolecat" %% "doobie-hikari"
+        "org.tpolecat" %% "doobie-hikari",
+        "org.tpolecat" %% "doobie-scalatest"
```
[the remark this answers is not in the document]
Comment (eryshev) [reply build.sbt]: ok

File: core/src/main/scala/com/criteo/cuttle/Executor.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (eryshev) [reply core/src/main/scala/com/criteo/cuttle/Executor.scala]: Indeed it wasn't the case, you could check updated PR

File: core/src/main/scala/com/criteo/cuttle/Executor.scala [hunk 4]
```diff
@@ -128,6 +128,17 @@ class CancellationListener private[cuttle] (execution: Execution[_], private[cut
...
+                                                    user: User,
+                                                    date: Instant,
+                                                    executions: Map[Execution[S], Promise[Completed]]) {
+  def toPausedJob(): PausedJob = PausedJob(id, user, date)
+}
+case class PausedJob(id: String, user: User, date: Instant) {
```
[the remark this answers is not in the document]
Comment (eryshev) [reply core/src/main/scala/com/criteo/cuttle/Executor.scala]: yep

File: core/src/main/scala/com/criteo/cuttle/Executor.scala [hunk 5]
```diff
@@ -820,13 +852,8 @@ class Executor[S <: Scheduling] private[cuttle] (val platforms: Seq[ExecutionPla
...
                     val cancelNow = atomic { implicit tx =>
-                      val isStillPaused = pausedState.get(job.id).getOrElse(Map.empty).contains(execution)
-                      if (isStillPaused) {
-                        val pausedExecutions = pausedState.get(job.id).getOrElse(Map.empty).filterKeys(_ == execution)
-                        pausedState += (job.id -> pausedExecutions)
-                        true
-                      } else {
-                        false
+                      pausedState.get(job.id).exists {
+                        case PausedJobWithExecutions(_, _, _, executions) if executions.contains(execution) => true
```
Comment (dufrannea) [inline core/src/main/scala/com/criteo/cuttle/Executor.scala]: pausedState += (job.id -> pausedExecutions) disappeared in the refacto

File: core/src/main/scala/com/criteo/cuttle/Executor.scala [hunk 6]
```diff
@@ -818,15 +853,17 @@ class Executor[S <: Scheduling] private[cuttle] (val platforms: Seq[ExecutionPla
...
                 case Paused =>
                   execution.streams.debug(s"Delayed because job ${execution.job.id} is paused")
+                  // we attach this callback to freshly created "Paused" execution
                   execution.onCancel { () =>
                     val cancelNow = atomic { implicit tx =>
-                      val isStillPaused = pausedState.get(job.id).getOrElse(Map.empty).contains(execution)
-                      if (isStillPaused) {
-                        val pausedExecutions = pausedState.get(job.id).getOrElse(Map.empty).filterKeys(_ == execution)
-                        pausedState += (job.id -> pausedExecutions)
+                      val job2Execution = pausedState.get(job.id).collectFirst {
```
Comment (dufrannea) [inline core/src/main/scala/com/criteo/cuttle/Executor.scala]: Please add a comment, and possibly rename job2Execution to something a bit more self-explanatory

File: core/src/main/scala/com/criteo/cuttle/App.scala [hunk 7]
```diff
@@ -11,6 +11,7 @@ import cats.implicits._
 import fs2.Stream
 import io.circe._
 import io.circe.syntax._
+import io.circe.generic.auto._
```
[the remark this answers is not in the document]
Comment (eryshev) [reply core/src/main/scala/com/criteo/cuttle/App.scala]: done

File: core/src/main/scala/com/criteo/cuttle/Executor.scala (hunk 6 above)
Comment (eryshev) [reply core/src/main/scala/com/criteo/cuttle/Executor.scala]: done

File: core/src/main/javascript/app/pages/Jobs.js [hunk 8]
```diff
@@ -163,14 +202,18 @@ class JobsComp extends React.Component<any, Props, State> {
...
     const setOfSelectedJobs = new Set(selectedJobs);
-
     const Data = () => {
       if (data && data.length && pausedJobs) {
-        const preparedData = data
-          .map(datum => ({
-            ...datum,
-            status: pausedJobs.has(datum.id) ? "paused" : "active"
-          }))
+        const jobs = data.map((job: Job) =>
```
[the remark this answers is not in the document]
Comment (eryshev) [reply core/src/main/javascript/app/pages/Jobs.js]: done

**Pull request merged**